### PR TITLE
resolves #800: Add WaitForState support for resource_gitlab_project branch protection

### LINKED ISSUE #800: Flaky test TestAccGitlabProject_initializeWithReadme
The `TestAccGitlabProject_initializeWithReadme` test has been failing on and off.

```
=== RUN   TestAccGitlabProject_initializeWithReadme
    resource_gitlab_project_test.go:328: Step 1/1 error: Check failed: Check 2/3 error: expected default branch to be protected
```

https://github.com/gitlabhq/terraform-provider-gitlab/runs/4928564294

### PR DESCRIPTION
**Pull request by PatrickRice-KSC**
This fixes an issue where branch protection ran in the background when creating a project, causing the default branch to occasionally not be protected.

## Description

Closes #800 
There was a race condition apparent in project creation where occasionally the branch protection would return `false` even though the branch had been protected asynchronously. This could cause issues if downstream resources used that output value. This also caused several tests that checked the branch protection status to be flakey and seemingly appear to fail rendomly when they were, in fact, failing properly.

## PR Checklist

- [x] Resource attributes match 1:1 the names and structure of the API resource in [the GitLab API documentation](https://docs.gitlab.com/ee/api/).
- [x] [Examples](/examples) are updated with:
    - A \*.tf file for the resource/s with at least one usage example
    - A \*.sh file for the resource/s with an import example (if applicable)
- [x] New resources have at minimum a basic test with three steps:
    - Create the resource
    - Update the attributes
    - Import the resource
- [x] No new `//lintignore` comments that came from copied code. Linter rules are meant to be enforced on new code.

### CODE REVIEW

**Comment by PatrickRice-KSC** [comment]
OK, well this is embarrassing given I'd run those tests locally. Investigating.

**Comment by PatrickRice-KSC** [comment]
OK, note to self. Don't forget when running make testacc locally to remove the limiter you put on it to only run your specific tests. Whoops.

The issue here was that if you don't set `default_branch` or `initialize_with_readme`, then the default branch obviously throws a 404 when you attempt to query it. Similar to other situations where this happens, I now catch the 404 status and exit the WaitForState loop.

I'm oddly still getting one failing test with this change, related to a data source for an archived project not returning an empty state. I wouldn't have changed that behavior I'd think, but I'm going to go ahead and commit the changes for tonight and keep looking at it tomorrow.

File: internal/provider/resource_gitlab_project.go [hunk 1]
```diff
@@ -620,6 +621,35 @@ func resourceGitlabProjectCreate(ctx context.Context, d *schema.ResourceData, me
...
+	// before we continue. Note this check should only be required when there is a custom default branch set
+	// See issue 800: https://github.com/gitlabhq/terraform-provider-gitlab/issues/800
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"false"},
+		Target:  []string{"true"},
+		Timeout: 2 * time.Minute, //The async action usually completes very quickly, within seconds. Don't wait too long.
+		Refresh: func() (interface{}, string, error) {
+			branch, response, err := client.Branches.GetBranch(project.ID, project.DefaultBranch, gitlab.WithContext(ctx))
+			if err != nil {
+				if response.StatusCode == 404 {
```
Comment (armsnyder) [inline internal/provider/resource_gitlab_project.go:634]: `response` may be nil here. Prefer to use the `is404(err)` helper function.

**Review by armsnyder — DISMISSED** [review]

File: internal/provider/resource_gitlab_project.go (hunk 1 above)
Comment (PatrickRice-KSC) [reply internal/provider/resource_gitlab_project.go:634]: There's one other reference using similar syntax as this one in the project resource, I'll go ahead and fix both of them tomorrow so we're consistent!

Comment (armsnyder) [reply internal/provider/resource_gitlab_project.go:634]: Thanks!

**Comment by timofurrer** [comment]
@PatrickRice-KSC thanks for the deep-dive into what caused this and thanks for the fix! Much appreciated 🎉

**Pull request merged**